### ISSUE TITLE
Rescuing typed exceptions with the suffix form

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#11017]
The code:

```crystal
1 rescue Exception
2 rescue ex : Exception
```

expands to:

```crystal
begin
  1
rescue
  Exception
end

begin
  2
rescue
  ex : Exception # this declaration has type `Nil` when used as an expression
end
```

1 is the same in Ruby, and I am not sure if anything can be done about it. But 2 is potentially misleading, because `ex` is never assigned but still accessible in the current scope, and thus it is also subject to #9882. Here are two possible solutions:

* Make 2 a syntax error. Note that Ruby disallows `2 rescue Exception => e`.
* Expand 2 instead to the following whenever the rescue expression is a `TypeDeclaration`:
  ```crystal
  begin
    2
  rescue ex : Exception
    ex
  end
  ```
  There would still be a syntax error if the `TypeDeclaration` has an initializer or if the type doesn't resolve to `Exception` or a subtype of it.

**Comment by Sija** [comment]
Syntax error sounds great considering it's an unintended behavior.

**Comment by asterite** [comment]
Yes, 2 could be changed into a syntax error. I was going to say that's it's probably not needed, because the docs are clear about `exp rescue exp` not allowing an exception type, or even capturing the exception, but... it seems this isn't even documented: https://crystal-lang.org/reference/syntax_and_semantics/exception_handling.html

**Comment by straight-shoota** [comment]
Maybe the compiler should also error when the rescue expression is a constant ending with `Error` or `Exception` (or, if it is any type / a subclass of Exception). It's just too easy to shoot yourself in the foot with that.

Alternatively, we could do some magic and transform to the intended semantics if the constant is a type.

**Comment by asterite** [comment]
Maybe we can just deprecate the suffix form, showing a warning message saying that it will be removed in 2.0

**Comment by straight-shoota** [comment]
`rescue nil` and other fallback values is pretty popular, though.

Btw. this is even used wrongly in stdlib:
https://github.com/crystal-lang/crystal/blob/fc0617ff701d328cab868b390750d492bb1f0ab6/src/http/server.cr#L517

**Comment by jhass** [comment]
Maybe it "just" needs to change syntax?

```cr
expr rescue = body
expr rescue var = body
expr rescue Type = body
expr rescue e : Type = body

# Invalid
expr rescue =
expr rescue
expr rescue body
expr rescue Type
expr rescue Type body
expr rescue var body
expr rescue var : Type
```

Not totally sure I like this, just thinking out loud. Road ahead could be to allow the new syntax, deprecate the old,  remove the old in 2.0.

**Comment by straight-shoota** [comment]
I don't think we need a new syntax for this. It's not Ver intuitive either.

I have the Impression that the naturally expected semantics are pretty solid actually. We just have to match the compiler's behaviour to that.

```
expr rescue body # => body
expr rescue # =
expr rescue ExceptionType
expr rescue OtherType # Error
expr rescue CONST
```

**Comment by jhass** [comment]
I really can't say I like the semantics of depending on the value of the constant. What happens for `FOO = ArgumentError; foo rescue FOO`?

**Comment by straight-shoota** [comment]
`Foo` is just a constant referring to a type, but not a type itself. So it would be treated as a value.

`alias FOO = ArgumentError` would of course would mean `FOO` is treated as a type.

**Comment by jhass** [comment]
Sorry, I'm not convinced. Whether a constant refers to a type, is assigned to a type or aliased to a type is not visible in its usage. As such semantically different behavior that is anything but raising an error in one case is just too subtle and brittle for my taste. Looking at `bar rescue Foo` should tell me what it's doing without having to chase up how `Foo` is defined.

**Comment by straight-shoota** [comment]
Yes, such behaviour that can't simply be reasoned about without context information is not ideal.

Maybe it's a bit of an excuse that assigning types to a constant (without being an alias) is very uncommon... 🤔

I suppose there would be a way to remove ambiguity if using a constant that is not a type would be invalid. There are obviously valid use cases for using a constant value as a fallback. There are of course workarounds: assign the constant to a local variable, make it accessible through a method or wrap the constant name in parenthesis.

**Comment by nobodywasishere** [comment]
An ameba rule was added for this recently https://github.com/crystal-ameba/ameba/pull/512
